lms/services/lti_h.py

```diff
@@ -91,33 +91,17 @@ def upsert_course_group(self):
         isn't allowed to create groups (e.g. if they're just a student) then
         show an error page instead of continuing with the LTI launch.
         """
-
-        self._upsert_h_group(
-            group_id=self._context.h_groupid,
-            group_name=self._context.h_group_name,
-            creator=self._context.h_user,
-        )
-
-        self.group_info_service.upsert(
-            authority_provided_id=self._context.h_authority_provided_id,
-            consumer_key=self._request.lti_user.oauth_consumer_key,
-            params=self._request.params,
-        )
-
-    def _upsert_h_group(self, group_id, group_name, creator):
-        """Update the group and create it if the user is an instructor."""
+        group_id = self._context.h_groupid
+        group_name = self._context.h_group_name
 
         try:
             self.h_api.update_group(group_id=group_id, group_name=group_name)
-            return
-
         except HAPINotFoundError:
             # The group hasn't been created in h yet.
+            self.h_api.create_group(group_id=group_id, group_name=group_name)
```

Comment:
```suggestion
            self.h_api.upsert_group(group_id=group_id, group_name=group_name)
```

This is the part that the upsert makes very weird now. We first try to update the group, and if that fails we then upsert it. I don't think we need the first call at all anymore.

Reply:
Hmm. Actually I think this might have made sense before. The logic was:

1. Whether the user is a teacher or a student, try to update the group
2. If the update fails:
   1. If the user is a student show an error page, since students couldn't create (or upsert) groups
   2. If the user is a teacher upsert the group

The user of upsert rather than update in the final step looks odd but upsert does not have the concurrency issues that create has so in fact it made sense to use upsert rather than create here.

And it _did_ have to be update and then upsert, because of the student vs teacher issue. 

It could have been done differently, and this would have been less confusing: for a student it could have been update then show error page if it fails, whereas for a teacher it could have just gone straight to update. I think the reason it ended up like it is, is probably because it originally used _create_ for the teacher because there was no upsert, and then that was changed to upsert later :man_shrugging: 

The naming of the `HAPI` method as `create_group()` when it actually does an upsert is wrong. But that name was introduced in a recent refactoring that I think probably got confused by the above and got the name wrong: https://github.com/hypothesis/lms/commit/d07495be2d1ab12f9281a74e5dcfad5eb0355ec4

Reply:
I want to minimise unnecessary changes to `LTIHService` and `HAPI` since they're scheduled to be completely changed to use the new bulk API ASAP. So any changes in the meantime just seem like unnecessary risk. Nonetheless, now that we're using the machine user it does seem to make sense just to call the upsert method right away so I'll do that

Reply:
I'm pretty sure I introduced the wrong naming here, as I had to reverse engineer what all these actions were doing when they were un-named. Win some lose some.

I'm not 100% sure if we're on the same page here, but I understood why it used to make sense, but I'm saying it no longer does.

Reply:
Ok:

* Renamed create_group() to upsert_group(): https://github.com/hypothesis/lms/pull/1614/commits/c589abb911a4c6e4143fec29eb6d854060b2738f
* Simplified the now-unnecessary update-then-upsert logic to just upsert: https://github.com/hypothesis/lms/pull/1614/commits/8ab3a4389940920f8f19610f129f4615e94cd215
* Re-tested and it still seems to work fine